### PR TITLE
Add 'density' kwarg to histogram

### LINKED ISSUE #7364: Histogram compatibility with numpy
In the documentation of numpy's histogram: https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram.html It is specified the keyword _normed_ will be removed in V2.0 and be replaced with _density_ (due to buggy behavior).

Should matplotlib be compatible with this naming, and atleast accept the keywoard _density_?

### PR DESCRIPTION
**Pull request by dstansby**
This is a merge of #7856 and #8408, both of which helped to introduce the `density` kwarg to `histogram`. Will definitely want a squash when it goes in as there were lots of little tricky bits to fix.

Fixes #7364

### CODE REVIEW

**Comment by tacaswell** [comment]
https://codecov.io/gh/matplotlib/matplotlib/compare/6582a9dfd7a0076e1ee3efab58f21545facf6107...2b3db003e7d9f2cc092dd140b0bceb94858df7db/src/lib/matplotlib/tests/test_axes.py#L2779 <- there is a test with a double name.

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -2776,6 +2776,47 @@ def test_hist_stacked_normed():
...
+        ax.hist((d1, d2), stacked=True, normed=False, density=True)
+
+    with pytest.raises(Exception):
+        ax.hist((d1, d2), stacked=True, normed=False, density=False)
+
+    with pytest.raises(Exception):
+        ax.hist((d1, d2), stacked=True, normed=True, density=False)
+
+
+def test_hist_stacked_density():
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py]: This test needs a different name.

Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: This test was being covered by one of the other added ones anyway, so I've just got rid of it.

**Comment by tacaswell** [comment]
Adding strict coverage changes on the _tests_ was a brilliant idea (not sure who suggested that, but I owe them a drink of their choice).

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -5879,55 +5878,61 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
+            times the bin width and not dividing by the total number
+            of observations. If *stacked* is also ``True``, the sum of the
             histograms is normalized to 1.
 
-            Default is ``False``
+            Default is ``None`` for both *normed* and *density*. If either is
+            set, then that value will be used. If neither are set, then the
+            args will be treated as ``False``.
+
+            If both are set to different things, the hist function raises an
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: should be 'if both are set'

**Comment by tacaswell** [comment]
@dstansby Can you squash this to three commits (one each for the code from @chelseatroy  and @moboehle and a third of your changes)?

**Comment by dstansby** [comment]
I think that's worked...

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -5879,55 +5878,60 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
 
             Default is ``None``
 
-        normed : boolean, optional
-            If `True`, the first element of the return tuple will
+        density : boolean, optional
+            If ``True``, the first element of the return tuple will
             be the counts normalized to form a probability density, i.e.,
             the area (or integral) under the histogram will sum to 1.
             This is achieved dividing the count by the number of observations
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: This is achieved by dividing the

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -6120,11 +6132,11 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
                     xmin = min(xmin, xi.min())
                     xmax = max(xmax, xi.max())
             bin_range = (xmin, xmax)
-
-        # hist_kwargs = dict(range=range, normed=bool(normed))
-        # We will handle the normed kwarg within mpl until we
-        # get to the point of requiring numpy >= 1.5.
-        hist_kwargs = dict(range=bin_range)
+        density = bool(density) or bool(normed)
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: why do you need to cast it to bool?

Comment (dstansby) [inline lib/matplotlib/axes/_axes.py]: If `density`/`normed` is `None` then `bool(None)` evaluates to `False`.

Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: We don't strictly _need_ the cast to bool, but it also does not hurt.

Comment (story645) [inline lib/matplotlib/axes/_axes.py]: but doesn't if None also evaluate to false when it's not an equality operation?

**Comment by dstansby** [comment]
Going to merge since there's two approvals

**Pull request merged**